runtime/master/src/main/java/edu/snu/nemo/runtime/master/BlockManagerMaster.java

```diff
@@ -368,4 +357,82 @@ public void onMessageWithContext(final ControlMessage.Message message, final Mes
       }
     }
   }
+
+  /**
+   * The handler of block location requests.
+   */
+  @VisibleForTesting
+  public static final class BlockLocationRequestHandler {
+    private final String blockId;
+    private final CompletableFuture<String> locationFuture;
+
+    /**
+     * Constructor.
+     *
+     * @param blockId the ID of the block.
+     */
+    BlockLocationRequestHandler(final String blockId) {
+      this.blockId = blockId;
+      this.locationFuture = new CompletableFuture<>();
+    }
+
+    /**
+     * Completes the block location future.
+     * If there is any pending request, replies with the completed location.
+     *
+     * @param location the location of the block.
+     */
+    void complete(final String location) {
+      locationFuture.complete(location);
+    }
+
+    /**
+     * Completes the block location future with failure.
+     * If there is any pending request, replies with the cause.
+     *
+     * @param throwable the cause of failure.
+     */
+    void completeExceptionally(final Throwable throwable) {
+      locationFuture.completeExceptionally(throwable);
+    }
+
+    /**
+     * Registers a request for the block location.
+     * If the location is already known, reply the location instantly.
+     *
+     * @param requestId      the ID of the block location request.
+     * @param messageContext the message context to reply.
+     */
+    void registerRequest(final long requestId,
+                         final MessageContext messageContext) {
+      final ControlMessage.BlockLocationInfoMsg.Builder infoMsgBuilder =
+          ControlMessage.BlockLocationInfoMsg.newBuilder()
+              .setRequestId(requestId)
+              .setBlockId(blockId);
+
+      locationFuture.whenComplete((location, throwable) -> {
+        if (throwable == null) {
+          infoMsgBuilder.setOwnerExecutorId(location);
+        } else {
+          infoMsgBuilder.setState(
+              convertBlockState(((AbsentBlockException) throwable).getState()));
+        }
+        messageContext.reply(
+            ControlMessage.Message.newBuilder()
+                .setId(RuntimeIdGenerator.generateMessageId())
+                .setListenerId(MessageEnvironment.EXECUTOR_MESSAGE_LISTENER_ID)
+                .setType(ControlMessage.MessageType.BlockLocationInfo)
+                .setBlockLocationInfoMsg(infoMsgBuilder.build())
+                .build());
+      });
+    }
+
+    /**
+     * @return the future of the block location.
+     */
+    @VisibleForTesting
+    public synchronized Future<String> getLocationFuture() {
```

Comment:
Remove? I think this method is only synchronizing with itself, which is unnecessary as it always returns the same object.